### PR TITLE
Fix LLVM compilation issues

### PR DESCRIPTION
**Pull request by jonmeow**
We've been having issues with asan builds on linux. This change should fix all of that. A build run can be found at:
https://github.com/carbon-language/carbon-lang/actions/runs/3093378863/jobs/5005683059

(currently in progress, but I'm expecting it to succeed at this point)

It may be that the issues with asan builds were actually related to caching. That is, maybe the brew build command didn't change enough between v14 and v15 that the cache hits were still an issue. We did notice this with 15.0.0 versus 15.0.1 include paths (that is, bazel wasn't happy using the cached results of a 15.0.0 build due to the skew in include paths). In order to address this, I've added CACHE_VERSION to the remote_cache setup. I've also set up corresponding buckets in Cloud.

However, I'm also switching Linux to llvm-15 and apt. I'd originally been looking at this because the issues were linux-specific, and we've previously had linux-specific issues with Homebrew. Although it may have been the cache all along, I would prefer to keep this setup (if nothing else, it made the caching issues more obvious, even though we were still confused by the include path manifestation).

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -59,33 +59,36 @@ jobs:
...
-        # Use llvm@14 because llvm (15) has sanitizer issues.
         run: |
           echo '*** Updating brew'
           brew update
           echo '*** Installing LLVM deps'
-          brew install --force-bottle --only-dependencies llvm@14
+          brew install --force-bottle --only-dependencies llvm
           echo '*** Installing LLVM itself'
-          brew install --force-bottle --force --verbose llvm@14
+          brew install --force-bottle --force --verbose llvm
```
Comment (zygoloid) [inline .github/workflows/tests.yaml:74]: I think we should explicitly force a version here, given that version changes require us to bump the cache version.

Should we / can we specify 15.0.1, or is @15 the best we can do?
Same question for the `llvm.sh` use below: can we force 15.0.1?

**Review by zygoloid — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (jonmeow) [reply .github/workflows/tests.yaml:74]: brew doesn't support `@15`, only stable: https://formulae.brew.sh/formula/llvm

llvm.sh only supports major versions.

**Comment by jonmeow** [comment]
Per discussion, adding a version to cache as a -D.
https://github.com/carbon-language/carbon-lang/actions/runs/3093378863/jobs/5005682813

File: bazel/cc_toolchains/clang_configuration.bzl [hunk 2]
```diff
@@ -18,25 +18,34 @@ def _run(repository_ctx, cmd):
...
-        return None
+    # Find a dot to indicate something like 'clang version 14.0.6', and grab the
+    # major version.
+    version_dot = version_output.find(".", version_start)
+    if version_dot != -1 and version_dot < version_newline:
+        clang_version = int(version_output[version_start:version_dot])
 
-    # Return the version as int.
-    return int(version_output[version_start:version_dot])
+    return (clang_version, clang_version_for_cache)
```
Comment (zygoloid) [inline bazel/cc_toolchains/clang_configuration.bzl:48]:
```suggestion
    return (clang_version, clang_version_for_cache.replace('"', '_').replace('\\', '_'))
```
May as well escape this here.

File: bazel/cc_toolchains/clang_configuration.bzl [hunk 3]
```diff
@@ -190,6 +202,7 @@ def _configure_clang_toolchain_impl(repository_ctx):
             "{LLVM_BINDIR}": str(arpath.dirname),
             "{CLANG_BINDIR}": str(clang.dirname),
             "{CLANG_VERSION}": str(clang_version),
+            "{CLANG_VERSION_FOR_CACHE}": clang_version_for_cache,
```
Comment (zygoloid) [inline bazel/cc_toolchains/clang_configuration.bzl]:
```suggestion
            "{CLANG_VERSION_FOR_CACHE}": clang_version_for_cache.replace('"', '_').replace('\\', '_'),
```

**Review by zygoloid — APPROVED** [review]

**Pull request merged**